pandas/core/frame.py

```diff
@@ -3973,10 +3973,25 @@ def set_index(self, keys, drop=True, append=False, inplace=False,
         -------
         dataframe : DataFrame
         """
-        inplace = validate_bool_kwarg(inplace, 'inplace')
+        from pandas import Series
+
         if not isinstance(keys, list):
             keys = [keys]
 
+        col_labels = [x for x in keys
```

Comment:
I wrote that:
> I also added better warnings in case there are duplicate column labels in keys, and some corresponding tests.

Currently, `df.set_index(['A', 'A'])` yields something like:
```ValueError: Duplicated level name: "A", assigned to level 1, is already used for level 0.```
which was
- untested
- not a very clear warning
- bad for perf, because all arrays get processed already and only upon creation of the `MultiIndex` is the error raised.

I fixed that with the code block above ('better warnings" is also in the title).